### ISSUE TITLE
T27: Run quality-biased additive best-first development cells

### ISSUE THREAD

**Issue by Sino-Huang** [Sino-Huang/multimodality_on_planning#65]
> *This was generated by AI during triage and revised for the issue-64 VLM-feasible v3 corpus.*

**Parent:** #38
**Plan ticket:** T27

## What to build
Run the quality-biased additive best-first development conditions for `best_first_add_w3`: pretrained base, process-SFT, random-valid, and exact-reference on the matched issue-64 v3 development panel.

The declared scalar priority is `g + 3*h_add` with canonical generation-serial tie-breaking and cheaper-path reopening. Plan optimality is not claimed or required.

## Acceptance criteria
- All four development conditions use the same issue-64 v3 selected development tasks and complete with semantically replayed compact episode evidence.
- Each selected exact reference contains at most 1,024 decisions; the frozen 2x allowance permits at most 2,048 VLM calls per episode.
- Qualify actual VLM throughput and wall-clock feasibility before rollout; issue a governed stop instead of silently reducing or segmenting an authorized episode.
- Run exactly one training seed (`17`) for each distinct model/cell configuration; rollout/reference seeds do not imply training-seed replication.
- The active path uses no hashes, checksums, artifact comparisons, regeneration comparisons, or integrity-check mode. Checks are limited to scientific episode semantics, algorithm invariants, coverage, and resource limits.
- No run starts without the matching authorization and gate receipt; outcomes distinguish `PASS`, `VALID_STOP`, `INVALID`, and `ANCESTOR_STOP`, and only `PASS` is scientific completion.

## Dependencies
#52 and #64 are complete.

---
**Governed vocabulary reference:** This ticket uses the governed vocabulary (operational competence, structural/process competence, validity, stop outcome, text-state, visual-state, multimodal-state) exactly as defined in `CONTEXT.md` section "Governed downstream vocabulary", governed by issue #39 (T01). Terms are reused verbatim; no local redefinition applies.

**Comment by Sino-Huang** [comment]
Apply the resource-bounded evaluation protocol established by #54:

- drive incremental episode sessions in deterministic rounds and use batched Transformers generation with one shared backbone per GPU, adapter switching, adapter-isolated caching, and at most one request per active episode per round;
- freeze a 2x matching-reference decision-call limit separately from expansion limits; exhaustion is a valid unsuccessful episode;
- evaluate only final checkpoints across five seeds, with base, learned, random-valid, and exact-reference conditions on the identical selected tasks;
- qualify throughput without success outcomes before rollout, using the actual precision/batch settings and a cost-qualified fallback panel if full coverage cannot fit;
- stop before new launches at the wall-clock cutoff, preserve completed episodes, replay every completed episode, and never let partial coverage satisfy the gate.

Report invariant-valid success, invalid-operation rate, budget usage, and learned-vs-best-control gain separately. #54 reached 100% learned success on its reduced panel but still correctly `VALID_STOP` because random-valid also reached 100%; avoid interpreting panel success alone as structural advantage.

**Comment by Sino-Huang** [comment]
Supervisor time-budget clarification: this ticket must not require multiple independent training-seed replicates. For each distinct model/cell configuration, authorize exactly one training run using seed 17 and evaluate only the final checkpoint from that run.

Any existing five-seeds requirement must be interpreted as rollout/reference/evaluation seeds applied to the single trained checkpoint where repeated evaluation is still useful, not as five separately trained checkpoints. Do not estimate or claim training-seed variance, and do not substitute rollout-seed variation for it; use whole-problem-instance uncertainty where applicable.

If a frozen scientific criterion truly requires independent training-seed replication, issue the appropriate governed VALID_STOP or gated-scope receipt instead of launching additional training runs.

**Comment by Sino-Huang** [comment]
Replacement selection after the retained issue-62 optimal-A* `VALID_STOP`:

- Independent Fast Downward screening on all 25 fixed development tasks found `eager_wastar([add()],w=2)` solved 25/25 with 16,622 total expansions and aggregate plan cost 1,699; `eager_greedy([add()])` without preferred operators solved 25/25 with 9,542 total expansions and aggregate plan cost 2,098.
- LM-cut/landmark variants were rejected: weighted LM-cut timed out on the hard VisitAll task even with few expansions because evaluation itself was expensive.
- Intermediate additive weights 3 and 5 lay between the selected endpoints and did not add a distinct scientific condition.
- The production adapters are named `best_first_add_w2` and `best_first_add_greedy`. Both use one eager scalar-priority runtime and `h_add`; the only setting differences are the priority formula and the standard reopening rule.
- The compact trace contract persists one copy of task/state facts plus per-decision typed target, trusted runtime result, and exact-input digest. Independent replay reconstructs every authoritative candidate/pruning fact and byte-identical input. Full frontier arrays, repeated complete inputs, landmarks, preferred operators, multiple queues, lazy evaluation, and trace segmentation are excluded.

This is a deliberately suboptimal best-first study. It makes no optimal-plan or admissible-search claim.

**Comment by Sino-Huang** [comment]
Issue #64 is complete. This ticket is now bound to the 64-pair v3 feasibility panel: maximum 1,024 reference decisions and 2,048 allowed VLM calls per episode, with throughput qualification required before launch. No integrity mechanism is authorized.

**Comment by Sino-Huang** [comment]
Implementation is pushed in `6259981` (`feat(planning): add best-first w3 development runner`).

Implemented for contract `issue-65-best-first-add-w3-development-v1`:
- the 23 issue-64 v3 development tasks only, with 7,095 exact w3 decisions, observed maximum 918, and a separate 2x per-episode call allowance (maximum 1,836);
- one two-epoch process-SFT LoRA run, training seed 17 only, 9,398 w3 train rows, 7,095 w3 dev rows, and 588 optimizer steps;
- exact and five-seed random-valid references, plus five rollout seeds for base and the single learned checkpoint; deterministic cache sharing means 28,380 worst-case unique model generations across the two model conditions while every logical episode retains its own allowance;
- outcome-blind two-A100 float32 throughput qualification before training or rollout; only the complete 23-task panel can pass, otherwise qualification emits a terminal `VALID_STOP`;
- two cost-balanced GPU rollout shards, semantic episode replay, whole-instance bootstrap reporting, and PASS / VALID_STOP / INVALID / ANCESTOR_STOP vocabulary;
- terminal progress and ETA logs, including a heartbeat every 30 seconds during model loading and blocking generation batches;
- numbered resumable training attempts and resumable per-episode references/rollouts.

The active path creates no hashes, checksums, artifact comparisons, regeneration comparisons, or follow-up check mode. The prior BFS-only prompt in the shared batched Qwen adapter was also corrected by allowing the best-first corpus/live message builder to be supplied explicitly.

Validation completed without a learning or model-rollout run:
- actual local SFT projection materialized: 9,398 train / 7,095 dev rows (67 MB);
- all 23 tasks passed exact plus random-valid semantic episode execution/replay probes (46 episodes);
- focused checks: 16 passed;
- complete planning benchmark: 353 passed (one unrelated Albumentations update warning);
- Ruff, Black, and basedpyright pass;
- complete workflow dry-run passes.

Operator command:
```bash
source ~/cd_vlaplan
cd /data/scratch/projects/punim0478/sukaih/multimodality_on_planning
python scripts/run_best_first_issue65.py all --devices cuda:0 cuda:1 --reference-workers 8
```

The command qualifies first. If qualification cannot certify the complete product inside 15 hours, it records and adjudicates `VALID_STOP` without starting training. If it passes, it continues through references, training, rollout, and adjudication. Resume an interrupted pre-cutoff run by adding `--resume`. Detailed stage commands are in `docs/issue-65-best-first-add-w3-development.md`.

The ticket remains open pending the real qualification and, only if authorized by that result, the remaining experiment.

**Comment by Sino-Huang** [comment]
Issue #65 completed under contract `issue-65-best-first-add-w3-development-v1` with scientific `PASS`. Compact terminal evidence is retained in `3487579` at `data/best_first_paired_phase_v3/issue65-terminal/result.json`.

Completion evidence:
- outcome-blind two-A100 qualification certified the complete 23-task issue-64 v3 development panel; lower-95% throughput was 0.393304 calls/s and projected rollout time was 43,403.66 seconds, below the frozen 15-hour ceiling;
- the selected traces remained bounded at a maximum of 918 reference decisions and 1,836 calls per episode;
- exactly one training run was performed: seed 17, two epochs, 588/588 optimizer steps, final checkpoint `checkpoint-588`;
- references completed: 23 exact-reference and 115 random-valid episodes;
- model rollout completed: 115 pretrained-base and 115 process-SFT episodes across two balanced GPU shards;
- all 368 retained episodes were independently reconstructed through the trusted runtime with the same scientific inputs, transitions, and results;
- final gate receipt: `gate:issue-65-best-first-add-w3-development-v1:attempt-001`, `PASS`, `scientific_completion: true`.

Results:
- exact reference: 23/23 invariant-valid successes;
- pretrained base: 0/115; all 115 terminated after one deterministic invalid operation;
- process SFT: 115/115, zero invalid operations, 0.5 call-budget usage;
- random-valid: 115/115, zero invalid operations;
- learned gain over the best control: 0.0; paired whole-instance bootstrap lower bound: 0.0.

The bounded interpretation is therefore: process SFT completely repaired the pretrained bases operational failure and executed every development episode correctly, but it did not establish superiority over random-valid because that control also saturated the panel. This is a completed development result, not a positive learned-versus-best-control advantage claim.

The active run used scientific semantic validation only and introduced no hash, checksum, artifact-comparison, regeneration-comparison, or follow-up check mechanism.
